service/dap/server.go

```diff
@@ -1559,10 +1604,103 @@ func (s *Server) onReverseContinueRequest(request *dap.ReverseContinueRequest) {
-	s.sendNotYetImplementedErrorResponse(request.Request)
+func variableByName(v *proc.Variable, name string) (*proc.Variable, bool) {
+	if v == nil {
+		return nil, false
+	}
+	if v.Name == name {
+		return v, true
+	}
+
+	for _, child := range v.Children {
+		if child.Name == name {
+			return &child, true
+		}
+		// No need to look into the child's children - DAP clients are expected
+		// to provide the variable reference number, so the name is the name of
+		// the variable's direct child.
+	}
+	return nil, false
+}
+
+// nameToEvaluateName finds the variable of the specified name
+// from the given variable and its children, and returns its
+// evaluateName.
+func (s *Server) nameToEvaluateName(v *fullyQualifiedVariable, name string) (string, error) {
+	children, err := s.childrenOfVariable(v)
+	if err != nil {
+		return "", err
+	}
+	evaluateName := name
+	for _, c := range children {
+		if c.Name == name {
+			if c.EvaluateName != "" {
+				evaluateName = c.EvaluateName
+			}
+			break
+		}
+	}
+	return evaluateName, nil
+}
+
+// onSetVariableRequest handles 'setVariable' requests.
+func (s *Server) onSetVariableRequest(request *dap.SetVariableRequest) {
+	arg := request.Arguments
+
+	v, ok := s.variableHandles.get(arg.VariablesReference)
+	if !ok {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to lookup variable", fmt.Sprintf("unknown reference %d", request.Arguments.VariablesReference))
+		return
+	}
+	// We need to translate the arg.Name to its evaluateName if the name
+	// refers to a field or element of a variable.
+	// https://github.com/microsoft/vscode/issues/120774
+	// Look up the evaluateName.
+	evaluateName, err := s.nameToEvaluateName(v, arg.Name)
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", err.Error())
+		return
+	}
+	named, ok := variableByName(v.Variable, arg.Name)
+	if !ok {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to lookup variable", fmt.Sprintf("unknown reference %d", request.Arguments.VariablesReference))
+		return
+	}
+	// Assume the top-most frame & the current goroutine.
+	// It is not clear if DAP allows SetVariable requests
+	// for variables or memory location not visible from the scope.
+	// TODO: keep track of the frame for a variable and allow alternate
+	// frame, or send explicit error messages for unsupported frames.
+	goid, frame := -1, 0
+
+	switch named.Kind {
+	case reflect.String:
+		if _, err := s.doCall(goid, frame, fmt.Sprintf("%v=%v", evaluateName, arg.Value)); err != nil {
+			s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", err.Error())
+			return
+		}
+	default:
+		if err := s.debugger.SetVariableInScope(goid, frame, 0, evaluateName, arg.Value); err != nil {
+			s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", err.Error())
+			return
+		}
+		// Send StoppedEvent so the client retrieve threads/scopes again.
```

Comment:
I can swear I wrote a comment about this, but I don't see it anymore.

So in a similar fashion as with call assignments, I think resetting all threads and frames and sending a stopped event here might be an overkill. You might end up with a different stop reason because you pick it from the debugger and not based on doCommand code. You will have to wait to load all threads. Etc. Unfortunately here vscode doesn't help us and doesn't send a scopes request to refresh the variables. And ideally we do want to refresh all variables in the selected scope, both in the UI and in the handles map. And the reason why I say variables and not just this one variable is because ideally we want three things updated:
(1) the value of the thing you just set
(2) the value of any children that got reset because you reset the parent
(3) the inlined value within the value of the parent variable
I tried this with javascript. They update (1) and (2), but not (3). The best way to update all three would be to file an issue against vscode and tell them they need to send a scopes request like they do after evaluate requests. 

Settling for (1) and (2) is good enough for the time being. And I think the way to do that is to return new variables reference in the setVariable response. That's how you update the cached value. You just let the old one be (we can optimize this eventually and clear, but for now we just stop using its key), and you give the editor a new reference for the refreshed variable.

You should be able to test this with two pointers p1 and p2 of he same type and then set variable p1 to p2. What should we see? The expanded dereferenced value should not disappear and should change to whatever *p2 is. Does that work?